### PR TITLE
Feature/cldn 1637

### PR DESCRIPTION
**Pull request by cccs-mmrouet**
Changes for both task to upgrade to trino 400 and to run some of the pipeline jobs inside a container.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -66,97 +66,139 @@ RUN chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
...
-   && set -x \
+    && set -x \
+    && update-ca-certificates \
 # install packages:
     && apt-get update \
 # packages for awt:
     && apt-get install libxext6 libxrender1 libxtst6 libxi6 libfreetype6 -y \
 # packages for user convenience:
-    && apt-get install ca-certificates ca-certificates-java git bash-completion vim sudo -y \
+    && apt-get install ca-certificates ca-certificates-java git bash-completion vim sudo unzip zip sed apt-utils -y \
```
Comment (cccs-tom) [inline Dockerfile]: I know we will probably stop using this image for Trino, but I've been trying to keep this line identical to cccs-main (to minimize merge conflicts in the future). See the datahub branch for how I handled it there https://github.com/CybercentreCanada/projector-docker/blob/108a318caa5ce71b3691a3f22e021f8ef31a823e/Dockerfile#L87-L94

File: Dockerfile [hunk 2]
```diff
@@ -66,97 +66,139 @@ RUN chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
...
 # packages for IDEA (to disable warnings):
-    && apt-get install procps -y \
-# clean apt to reduce image size:
-    && rm -rf /var/lib/apt/lists/* \
-    && rm -rf /var/cache/apt
```
Comment (cccs-tom) [inline Dockerfile]: Why are we removing these cleaning steps? This will bloat the image for no reason.

File: Dockerfile [hunk 3]
```diff
@@ -66,97 +66,139 @@ RUN chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
...
     && if [ "${downloadUrl#*pycharm}" != "$downloadUrl" ]; then apt-get install python2 python3 python3-distutils python3-pip python3-setuptools -y; else echo "Not pycharm"; fi \
-    && if [ "${downloadUrl#*rider}" != "$downloadUrl" ]; then apt install apt-transport-https dirmngr gnupg ca-certificates -y && apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF && echo "deb https://download.mono-project.com/repo/debian stable-buster main" | tee /etc/apt/sources.list.d/mono-official-stable.list && apt update && apt install mono-devel -y && apt install wget -y && wget https://packages.microsoft.com/config/debian/10/packages-microsoft-prod.deb -O packages-microsoft-prod.deb && dpkg -i packages-microsoft-prod.deb && rm packages-microsoft-prod.deb && apt-get update && apt-get install -y apt-transport-https && apt-get update && apt-get install -y dotnet-sdk-3.1 aspnetcore-runtime-3.1; else echo "Not rider"; fi \
-# clean apt to reduce image size:
-    && rm -rf /var/lib/apt/lists/* \
-    && rm -rf /var/cache/apt
```
Comment (cccs-tom) [inline Dockerfile]: Same comment here - do not remove this cleanup

File: Dockerfile [hunk 4]
```diff
@@ -66,97 +66,139 @@ RUN chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
...
+# https://docs.github.com/en/github/authenticating-to-github/githubs-ssh-key-fingerprints
+RUN mkdir -p /home/${PROJECTOR_USER_NAME}/.ssh \
+    && echo "github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ==" >> /home/${PROJECTOR_USER_NAME}/.ssh/known_hosts \
+    && chown -R ${PROJECTOR_USER_NAME} /home/${PROJECTOR_USER_NAME}/.ssh \
+    && touch /usr/local/share/bash_history \
+    && chown ${PROJECTOR_USER_NAME} /usr/local/share/bash_history
+
+# Install additional OS packages.
+RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+    && apt-get -y install --no-install-recommends bash-completion vim \
```
Comment (cccs-tom) [inline Dockerfile]: These already got installed on line 84, did they not?

File: Dockerfile [hunk 5]
```diff
@@ -66,97 +66,139 @@ RUN chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
...
-    && usermod -a -G azure_pipelines_docker AzDevOps_azpcontainer \
+    &&  set -x \
+    && for cert in /usr/local/share/ca-certificates/*; do \
+        openssl x509 -outform der -in "$cert" -out /tmp/certificate.der; \
+        $PROJECTOR_DIR/ide/jbr/bin/keytool -import -alias "$cert" -keystore $PROJECTOR_DIR/ide/jbr/lib/security/cacerts -file /tmp/certificate.der -deststorepass changeit -noprompt; \
+    done \
+    && rm /tmp/certificate.der
 
+USER $PROJECTOR_USER_NAME:$PROJECTOR_USER_GID
+ENV HOME /home/$PROJECTOR_USER_NAME
```
Comment (cccs-tom) [inline Dockerfile:205]: This should be automatic. I don't think this is required...

File: static/library-scripts/azcli-debian.sh [hunk 6]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env bash
```
Comment (cccs-tom) [inline static/library-scripts/azcli-debian.sh:1]: Just double-checking... This file is taken as-is from the MS repo, correct? No modifications?

File: static/library-scripts/docker-in-docker-debian.sh [hunk 7]
```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env bash
```
Comment (cccs-tom) [inline static/library-scripts/docker-in-docker-debian.sh:1]: Just double-checking... This file is taken as-is from the MS repo, correct? No modifications?

**Review by cccs-tom — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (cccs-mmrouet) [reply Dockerfile]: je vais le rajouter

File: static/library-scripts/docker-in-docker-debian.sh (hunk 7 above)
Comment (cccs-mmrouet) [reply static/library-scripts/docker-in-docker-debian.sh:1]: I don't think I made changes to it indeed.

File: static/library-scripts/azcli-debian.sh (hunk 6 above)
Comment (cccs-mmrouet) [reply static/library-scripts/azcli-debian.sh:1]: Iirc yes

File: Dockerfile (hunk 5 above)
Comment (cccs-mmrouet) [reply Dockerfile:205]: It was in the devcontainer dockerfile so that's why I brought it over...

File: Dockerfile (hunk 4 above)
Comment (cccs-mmrouet) [reply Dockerfile]: not the bash-completion but vim yes. Will remove vim

File: Dockerfile (hunk 3 above)
Comment (cccs-mmrouet) [reply Dockerfile]: What if I do it once at the end?  Is isn't it enough?

Comment (cccs-mmrouet) [reply Dockerfile]: fixed

File: Dockerfile (hunk 1 above)
Comment (cccs-mmrouet) [reply Dockerfile]: done

File: Dockerfile (hunk 2 above)
Comment (cccs-mmrouet) [reply Dockerfile]: Done

File: Dockerfile (hunk 4 above)
Comment (cccs-mmrouet) [reply Dockerfile]: done

Comment (cccs-tom) [reply Dockerfile]: I believe your branch is out of date with the trino branch...
https://github.com/CybercentreCanada/projector-docker/blob/5b05c87b75a4aeaffa0d1bee5a94e9f11eca38bb/Dockerfile#L83

File: static/library-scripts/docker-in-docker-debian.sh (hunk 7 above)
Comment (cccs-mmrouet) [reply static/library-scripts/docker-in-docker-debian.sh:1]: confirmed, j'ai rien changé

File: static/library-scripts/azcli-debian.sh (hunk 6 above)
Comment (cccs-mmrouet) [reply static/library-scripts/azcli-debian.sh:1]: rien changé ici non plus

File: Dockerfile (hunk 5 above)
Comment (cccs-mmrouet) [reply Dockerfile:205]: I kept this one.

**Review by cccs-tom — APPROVED** [review]

File: Dockerfile [hunk 8]
```diff
@@ -66,24 +66,27 @@ RUN chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
...
     && apt-get update \
 # packages for awt:
     && apt-get install libxext6 libxrender1 libxtst6 libxi6 libfreetype6 -y \
 # packages for user convenience:
+    && apt-get install git bash-completion vim -y \
+ # packages and CA certs added for CCCS use (general use) 
+    && apt-get install ca-certificates jq vim -y \ 
+    && update-ca-certificates \ 
+    && apt-get install sudo unzip zip sed apt-utils -y \
     && apt-get install git bash-completion sudo -y \
```
Comment (cccs-nik) [inline Dockerfile]: This could be one line or at least one apt-get statement. Also sudo is there twice.

File: Dockerfile [hunk 9]
```diff
@@ -66,24 +66,27 @@ RUN chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
...
 ADD ${extraCaCertsDir} /usr/local/share/ca-certificates/
 
 RUN true \
 # Any command which returns non-zero exit code will cause this shell script to exit immediately:
-   && set -e \
+    && set -e \
 # Activate debugging to show execution details: all commands will be printed before execution
-   && set -x \
+    && set -x \
+    && update-ca-certificates \
```
Comment (cccs-nik) [inline Dockerfile]: You're already running this on line 87. I think this should be run after you install ca-certificates anyway?

File: Dockerfile [hunk 10]
```diff
@@ -111,33 +114,101 @@ ENV PROJECTOR_DIR /projector
...
+
+# The steps below are to set up Trino environment:
+ARG TRINO_VERSION="400"
+
+RUN true \
+# Any command which returns non-zero exit code will cause this shell script to exit immediately:
+    && set -e \
+# Activate debugging to show execution details: all commands will be printed before execution
+    && set -x \
+    && apt-get update  && apt-get install -y apt-transport-https \
```
Comment (cccs-nik) [inline Dockerfile]: This should be two separate lines

File: Dockerfile (hunk 8 above)
Comment (cccs-mmrouet) [reply Dockerfile]: Je vais ajuster.

File: Dockerfile (hunk 9 above)
Comment (cccs-mmrouet) [reply Dockerfile]: Done

File: Dockerfile (hunk 8 above)
Comment (cccs-mmrouet) [reply Dockerfile]: Done

File: Dockerfile (hunk 10 above)
Comment (cccs-mmrouet) [reply Dockerfile]: oups, c'est fait

**Review by cccs-nik — APPROVED** [review]

**Pull request merged**